gift_card.py

```diff
@@ -348,3 +352,141 @@ def parse(cls, report, records, data, localcontext):
+    amount = fields.Numeric('Amount', digits=(16, 2), required=True)
+
+
+class GiftCardRedeemDone(ModelView):
+    "Gift Card Redeem Done View"
+    __name__ = 'gift_card.redeem.end'
+
+
+class GiftCardRedeemWizard(Wizard):
+    "Gift Card Redeem Wizard"
+    __name__ = 'gift_card.redeem.wizard'
+
+    start = StateView(
+        'gift_card.redeem.start',
+        'gift_card.redeem_start_view_form',
+        [
+            Button('Cancel', 'end', 'tryton-cancel'),
+            Button(
+                'Redeem', 'redeem',
+                'tryton-ok', default=True
+            )
+        ]
+    )
+    redeem = StateTransition()
+    done = StateView(
+        'gift_card.redeem.end',
+        'gift_card.redeem_done_view_form',
+        [
+            Button('OK', 'end', 'tryton-ok')
+        ]
+    )
+
+    @classmethod
+    def __setup__(cls):
+        super(GiftCardRedeemWizard, cls).__setup__()
+        cls._error_messages.update({
+            'gift_card_inactive': (
+                'The gift card to be redeemed must be in active state.'
+            ),
+            'gift_card_redeemed': (
+                'This gift card has already been redeemed.'
+            ),
+            'multiple_giftcards': (
+                'You can only redeem a single gift card at a'
+                ' time.'
+            ),
+            'amount_exceeded': (
+                'The amount entered is greater than amount available in the'
+                ' gift card.'
+            ),
+        })
+
+    def default_start(self, data):
+        """
+        Initial state of redeem wizard.
+        """
+        Gateway = Pool().get('payment_gateway.gateway')
+        GiftCard = Pool().get('gift_card.gift_card')
+
+        gift_cards = GiftCard.browse(Transaction().context.get('active_ids'))
+        if len(gift_cards) > 1:  # pragma: no cover
+            self.raise_user_error('multiple_giftcards')
+        gift_card = gift_cards[0]
+        self.check_giftcard_state(gift_card)
+
+        gateways = Gateway.search([
+            ('method', '=', 'gift_card'),
+            ('active', '=', True),
+        ])
+        return {
+            'gift_card': gift_card.id,
+            'gateway': gateways and gateways[0].id or None,
+        }
+
+    def transition_redeem(self):
+        """
+        Transition where PaymentTransaction is created and associated
+        with current gift card.
+        """
+        GiftCard = Pool().get('gift_card.gift_card')
+        PaymentTransaction = Pool().get('payment_gateway.transaction')
+        Date = Pool().get('ir.date')
+
+        self.validate_amount(self.start.amount, self.start.gift_card)
+
+        transaction, = PaymentTransaction.create([{
+            'description': self.start.description,
+            'date': Date.today(),
+            'party': self.start.party.id,
+            'address': (
+                self.start.party.addresses and
+                self.start.party.addresses[0].id or None
```

Comment:
Don't use index 0 address. ask user for this and there you can set default to invoice address or index 0